lib/rspec/mocks/proxy.rb

```diff
@@ -407,23 +407,29 @@ class PartialClassDoubleProxy < PartialDoubleProxy
     class ProxyForNil < PartialDoubleProxy
       def initialize(order_group)
         @warn_about_expectations = true
+        @disallow_expectations_on_nil = false
         super(nil, order_group)
       end
 
       attr_accessor :warn_about_expectations
+      attr_accessor :disallow_expectations_on_nil
       alias warn_about_expectations? warn_about_expectations
+      alias disallow_expectations_on_nil? disallow_expectations_on_nil
 
       def add_message_expectation(method_name, opts={}, &block)
+        raise_error(method_name) if disallow_expectations_on_nil?
         warn(method_name) if warn_about_expectations?
         super
       end
 
       def add_negative_message_expectation(location, method_name, &implementation)
+        raise_error(method_name) if disallow_expectations_on_nil?
         warn(method_name) if warn_about_expectations?
```

Comment:
Can we collapse these into one if check for performance reasons? e.g:

```Ruby
...
warn_or_raise(method_name) if warn_or_disallow_expectations?
...

def warn_or_raise(method_name)
  if disallow_expectations?
    raise_error(method_name)
  else
    warn(method_name)
  end
end
```

We'd also need to refactor the accessors to make that work but it means that a happy path will only encounter 1 if check.